gcc/rust/resolve/rust-name-resolution-context.h

```diff
@@ -156,6 +156,108 @@ class Definition
   NodeId id;
 };
 
+struct Binding
+{
+  enum class Kind
+  {
+    Product,
+    Or,
+  } kind;
+
+  std::unordered_set<Identifier> set;
+
+  Binding (Binding::Kind kind) : kind (kind) {}
+};
+
+/**
+ * Used to identify the source of a binding, and emit the correct error message.
+ */
+enum class BindingSource
+{
+  Match,
+  Let,
+  For,
+  /* Closure param or function param */
+  Param
+};
+
+class BindingLayer
+{
+  BindingSource source;
+  std::vector<Binding> bindings;
+
+  bool bind_test (Identifier ident, Binding::Kind kind);
+
+public:
+  void push (Binding::Kind kind);
+
+  BindingLayer (BindingSource source);
+
+  /**
+   * Identifies if the identifier has been used in a product binding context.
+   * eg. `let (a, a) = test();`
+   */
+  bool and_binded (Identifier ident);
+
+  /**
+   * Identifies if the identifier has been used in a or context.
+   * eg. `let (a, 1) | (a, 2) = test()`
+   */
+  bool or_binded (Identifier ident);
+
+  void insert_ident (Identifier ident);
+
+  void merge ();
+
+  BindingSource get_source () const;
+};
+
+class BindingContext
+{
+  std::vector<BindingLayer> bindings;
+
+public:
+  /**
+   * Identifies if the identifier is and-binded in the current layer's bindings.
+   */
+  bool and_binded (Identifier ident);
+
+  /**
+   * Identifies if the identifier is or-binded in the current layer's bindings.
+   */
+  bool or_binded (Identifier ident);
```

Comment:
I would rename these `is_and_bound()` and `is_or_bound()` since they return a boolean. I was a bit confused when reading the code using these functions above